lib/api/src/sys/externals/memory.rs

```diff
@@ -72,47 +72,6 @@ impl Memory {
         self.handle.get(store.as_store_ref().objects()).ty()
     }
 
-    /// Returns the pointer to the raw bytes of the `Memory`.
-    //
-    // This used by wasmer-emscripten and wasmer-c-api, but should be treated
-    // as deprecated and not used in future code.
-    #[doc(hidden)]
-    pub fn data_ptr(&self, store: &impl AsStoreRef) -> *mut u8 {
-        self.buffer(store).base
-    }
-
-    /// Returns the size (in bytes) of the `Memory`.
-    pub fn data_size(&self, store: &impl AsStoreRef) -> u64 {
-        self.buffer(store).len.try_into().unwrap()
-    }
-
-    /// Retrieve a slice of the memory contents.
-    ///
-    /// # Safety
```

Comment:
We shouldn't modify the API of non-shared memory. We should either:
* create a new `SharedMemory` that is able to lock
* preserve the current memory API and depending if the memory is shared o no, to work on the MutexGuard or not

Reply:
SharedMemory would still require guards that abstract the access to the memory where none shared memory could be made to be a zero cost abstraction it will still mandate some changes in the way Wasmer API accesses memory. So the main difference becomes if they implement the lock inside or not. Also we would need to figure out a way to decide when to use SharedMemory and when not to use it. With WASI we don't know if its multithreaded or not upfront until its actually used (having import memory is not conclusive as one may wish to pass in memory but be singlethreaded). Hence in the WASI mode we would have to assume all programs are multithreaded and hence need SharedMemory. Beyond this there is no technical limitation to implementing what you say and can be made optional on the "enable_threads" argument

Reply:
Drastically simplified the changes around memory much closer to the original code. I had made an incorrect assumption which means the multithreading risks are much lower than I thought thus requiring much less protection. Wasmer uses mmap which means it does not change the base pointer address during grow operations, further it never shrinks in size... with those two constraints together it means there is no need to protect against a grow operation in parallel with normal memory operation, we only need to protect against concurrent grow operations which is only in one small part of the code. A simple RwLock around mmap is thus all that's needed.

Reply:
This is now implemented with VMSharedMemory instead and preserves the original API where possible